modules/commons/src/main/java/org/apache/ignite/internal/thread/context/ThreadContextData.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.thread.context;
+
+/** */
+class ThreadContextData {
+    /** */
+    private static final ScopedAttributeValueStack<?>[] EMPTY = new ScopedAttributeValueStack[0];
+
+    /** */
+    private int activeScopeDepth;
+
+    /** */
+    private int activeAttrsCnt;
+
+    /** */
+    private ScopedAttributeValueStack<?>[] attrs = EMPTY;
+
+    /** */
+    <T> T get(ThreadContextAttribute<T> attr) {
+        if (attr.id() >= attrs.length)
```

Comment:
Shoulud we throw assert instead? Is it correct to access to the attribute we actually didn't set?

Could it lead to an error, when user forget to set the var and received default value instead?

Reply:
No, we shouldn't. The initial value of an attribute literally means - the value returned by the ThreadContext#get method if the attribute's value is not explicitly set.

It can be used to skip explicitly setting attributes for some internal operations and still get some value for attribute.